static/js/components/widgets/HierarchicalSelectField.tsx

```diff
@@ -0,0 +1,168 @@
+import React, { SyntheticEvent, useCallback, useMemo, useState } from "react"
+import { sortedUniqBy, sortBy, times } from "lodash"
+
+import SelectField, { Option } from "./SelectField"
+
+export type Level = {
+  label: string
+  name: string
+}
+type OptionsMap = Record<string, Record<string, any> | string[]>
+type HierarchicalSelection = Array<string | null>
+type Props = {
+  name: string
+  levels: Level[] // eslint-disable-line camelcase
+  value: HierarchicalSelection[] | null
+  onChange: (event: any) => void
+  options_map: OptionsMap // eslint-disable-line camelcase
+}
+
+const asOption = (str: string | null) => ({
+  label: str ?? "-- empty --",
+  value: str ?? ""
+})
+
+const describeSelection = (selection: HierarchicalSelection): string =>
+  selection.filter(Boolean).join(" - ")
+
+/**
+ This function takes field data and the current selection and returns a set of options
+ to be displayed in each dropdown.
+*/
+export const calcOptions = (
+  optionsMap: OptionsMap,
+  selection: HierarchicalSelection,
+  levels: Level[]
+): Option[][] => {
+  // currentOptionsMap is the piece of optionsMap which is at the level depth being looked at.
+  // If falsey, assume there are no options to show.
+  let currentOptionsMap: any = optionsMap
+
+  // Iterate through the levels in order to show the right options.
+  // When an option is selected at one level, options in deeper levels are filtered on that selection.
+  return levels.map((_, levelIdx) => {
+    const selectedItem = selection[levelIdx]
+
+    // null indicates an empty selection, and we want to show that as the first option
+    let optionValues: Array<string | null> = [null]
+    if (!currentOptionsMap) {
+      // If the previous select field has no selection, then no select field at a deeper level
+      // should have options to select.
+    } else if (Array.isArray(currentOptionsMap)) {
+      optionValues = optionValues.concat(sortBy(currentOptionsMap))
+      // if currentOptionsMap is an array, we are at the deepest level
+      currentOptionsMap = null
+    } else {
+      console.log("level", levelIdx, selectedItem, currentOptionsMap)
```

Comment:
Is this debug logging, or did you mean to leave it in for some reason?

Reply:
Thanks, yes this was for debugging